File: repository/enderecoDAO.js
import conectar from "./Conexao.js";
import Endereco from "../model/endereco.js";

export default class EnderecoDAO {
    constructor() {
        this.init();
    }

    async init() {
        try {
            const conexao = await conectar.connect();
            const sql = `
            CREATE TABLE IF NOT EXISTS endereco (
                endereco_id SERIAL PRIMARY KEY,
                endereco_rua VARCHAR(255) NOT NULL,
                endereco_numero INT NOT NULL,
                endereco_complemento VARCHAR(255),
                endereco_cep VARCHAR(20) NOT NULL,
                endereco_uf VARCHAR(2) NOT NULL,
                endereco_cidade VARCHAR(100) NOT NULL
            );
            `;
            await conexao.query(sql);
            await conexao.release();
        } catch (erro) {
            console.log("Erro ao iniciar tabela endereco: " + erro.message);
        }
    }

    async gravar(endereco) {
        if (endereco instanceof Endereco) {
            const conexao = await conectar.connect();
            const sql = `INSERT INTO endereco (endereco_rua, endereco_numero, endereco_complemento, endereco_cep, endereco_uf, endereco_cidade) 
                VALUES ($1, $2, $3, $4, $5, $6) RETURNING endereco_id`;
            const parametros = [
                endereco.rua,
                endereco.numero,
                endereco.complemento,
                endereco.cep,
                endereco.uf,
                endereco.cidade
            ];
            const resultado = await conexao.query(sql, parametros);
            await conexao.release();
            return resultado.rows[0].endereco_id;
        }
    }

    async alterar(endereco) {
        if (endereco instanceof Endereco) {
            const conexao = await conectar.connect();
            const sql = `UPDATE endereco SET endereco_rua = $1, endereco_numero = $2, endereco_complemento = $3, endereco_cep = $4, endereco_uf = $5, endereco_cidade = $6 
                WHERE endereco_id = $7`;
            const parametros = [
                endereco.rua,
                endereco.numero,
                endereco.complemento,
                endereco.cep,
                endereco.uf,
                endereco.cidade,
                endereco.id
            ];
            await conexao.query(sql, parametros);
            await conexao.release();
        }
    }

    async apagar(id) {
        const conexao = await conectar.connect();
        const sql = `DELETE FROM endereco WHERE endereco_id = $1`;
        const parametros = [id];
        await conexao.query(sql, parametros);
        await conexao.release();
    }

    async get(filtro) {
        const conexao = await conectar.connect();
        let parametros;
        let sql;
        if (!filtro) {
            filtro = "";
        }
        if (!isNaN(filtro)) {
            sql = `SELECT * FROM endereco WHERE endereco_id = $1`;
            parametros = [filtro];
        } else {
            sql = `SELECT * FROM endereco WHERE endereco_rua LIKE $1 OR endereco_cidade LIKE $2`;
            parametros = [`%${filtro}%`, `%${filtro}%`];
        }
        const resultado = await conexao.query(sql, parametros);
        await conexao.release();
        const linhas = resultado.rows || [];
        if (linhas.length > 0) {
            let listaEnderecos = linhas.map(linha => new Endereco(
                linha.endereco_id,
                linha.endereco_rua,
                linha.endereco_numero,
                linha.endereco_complemento,
                linha.endereco_cep,
                linha.endereco_uf,
                linha.endereco_cidade
            ));
            return listaEnderecos.length === 1 ? listaEnderecos[0] : listaEnderecos;
        }
        return null;
    }
}

File: src/main/java/sistema/saga/sagabackend/repository/EnderecoDAO.java
package sistema.saga.sagabackend.repository;

import org.springframework.stereotype.Repository;
import sistema.saga.sagabackend.model.Endereco;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

@Repository
public class EnderecoDAO {

    public EnderecoDAO() {
    }

    public boolean gravar(Endereco endereco, Conexao conexao) {
        String sql;
        if(endereco.getComplemento()==null){
            sql = """
                INSERT INTO endereco(endereco_rua, endereco_numero, endereco_cep, endereco_cidade, endereco_uf) 
                VALUES ('#1', '#2', '#4', '#5', '#6')
                """;
        }
        else{
            sql = """
                INSERT INTO endereco(endereco_rua, endereco_numero, endereco_complemento, endereco_cep, endereco_cidade, endereco_uf) 
                VALUES ('#1', '#2', '#3', '#4', '#5', '#6')
                """;
            sql = sql.replace("#3", endereco.getComplemento());
        }
        sql = sql.replace("#1", endereco.getRua());
        sql = sql.replace("#2",""+endereco.getNumero());
        sql = sql.replace("#4", endereco.getCep());
        sql = sql.replace("#5", endereco.getCidade());
        sql = sql.replace("#6", endereco.getUf());

        return conexao.manipular(sql);
    }

    public boolean alterar(Endereco endereco, Conexao conexao) {
        String sql;
        if(endereco.getComplemento()==null){
            sql = """
                UPDATE endereco 
                SET endereco_rua = '#1', endereco_numero = '#2', endereco_cep = '#4', endereco_cidade = '#5', endereco_uf = '#6' 
                WHERE endereco_id = '#7'
                """;
        }
        else{
            sql = """
                UPDATE endereco 
                SET endereco_rua = '#1', endereco_numero = '#2', endereco_complemento = '#3', 
                endereco_cep = '#4', endereco_cidade = '#5', endereco_uf = '#6' 
                WHERE endereco_id = '#7'
                """;
            sql = sql.replace("#3", endereco.getComplemento());
        }
        sql = sql.replace("#1", endereco.getRua());
        sql = sql.replace("#2", ""+endereco.getNumero());
        sql = sql.replace("#4", endereco.getCep());
        sql = sql.replace("#5", endereco.getCidade());
        sql = sql.replace("#6", endereco.getUf());
        sql = sql.replace("#7", ""+endereco.getId());

        return conexao.manipular(sql);
    }

    public boolean apagar(Endereco endereco, Conexao conexao) {
        String sql = "DELETE FROM endereco WHERE endereco_id = '#1'";
        sql = sql.replace("#1", ""+endereco.getId());

        return conexao.manipular(sql);
    }

    public Endereco getEndereco(Endereco endereco, Conexao conexao) {
        String sql = "SELECT * FROM endereco WHERE endereco_rua like '#1' and endereco_numero = '#2'";
        sql = sql.replace("#1", endereco.getRua());
        sql = sql.replace("#2",""+ endereco.getNumero());
        try {
            ResultSet resultSet= conexao.consultar(sql);
            if (resultSet.next()) {
                return new Endereco(
                        resultSet.getLong("endereco_id"),
                        resultSet.getString("endereco_rua"),
                        resultSet.getInt("endereco_numero"),
                        resultSet.getString("endereco_complemento"),
                        resultSet.getString("endereco_cep"),
                        resultSet.getString("endereco_cidade"),
                        resultSet.getString("endereco_uf")
                );
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public Endereco getEndereco(int id, Conexao conexao) {
        String sql = "SELECT * FROM endereco WHERE endereco_id = '#1'";
        sql = sql.replace("#1", ""+id);
        try {
            ResultSet resultSet= conexao.consultar(sql);
            if (resultSet.next()) {
                String complemento= resultSet.getString("endereco_complemento");
                return new Endereco(
                        resultSet.getLong("endereco_id"),
                        resultSet.getString("endereco_rua"),
                        resultSet.getInt("endereco_numero"),
                        complemento,
                        resultSet.getString("endereco_cep"),
                        resultSet.getString("endereco_cidade"),
                        resultSet.getString("endereco_uf")
                );
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public List<Endereco> get(String filtro,Conexao conexao) {
        String sql = "SELECT * FROM endereco WHERE endereco_cidade LIKE '%#1%'";
        sql = sql.replace("#1", filtro);

        List<Endereco> enderecos = new ArrayList<>();
        try {
            ResultSet resultSet = conexao.consultar(sql);
            while (resultSet.next()) {
                enderecos.add(new Endereco(
                        resultSet.getLong("endereco_id"),
                        resultSet.getString("endereco_rua"),
                        resultSet.getInt("endereco_numero"),
                        resultSet.getString("endereco_complemento"),
                        resultSet.getString("endereco_cep"),
                        resultSet.getString("endereco_cidade"),
                        resultSet.getString("endereco_uf")
                ));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return enderecos;
    }
}
